src/Interpreters/ExpressionAnalyzer.cpp

```diff
@@ -1091,15 +1090,8 @@ ExpressionAnalysisResult::ExpressionAnalysisResult(
         {
             query_analyzer.appendJoinLeftKeys(chain, only_types || !first_stage);
 
-            before_join = chain.getLastActions(true);
-            if (before_join)
-                chain.addStep();
-
-            query_analyzer.appendJoin(chain);
-
-            join = chain.getLastActions();
-            if (!join)
-                throw Exception("No expected JOIN", ErrorCodes::LOGICAL_ERROR);
+            before_join = chain.getLastActions();
+            join = query_analyzer.appendJoin(chain);
             chain.addStep();
```

Comment:
Do we need one more step here? It looksl like we've already added one in `appendJoin()`.

Reply:
I suppose that we need, cause in following code we expect the last step in chain is Expression